src/rtcModule/webrtc.cpp

```diff
@@ -0,0 +1,2056 @@
+
+RtcModule::RtcModule(karere::Client& client, IGlobalHandler& handler,
+  IRtcCrypto* crypto, const char* iceServers)
+: IRtcModule(client, handler, crypto, crypto->anonymizeId(client.myHandle())),
+  mTurnServerProvider(client.api, "turn", iceServers, 3600)
+{
+    if (!artc::isInitialized())
+    {
+        artc::init(nullptr);
+        RTCM_LOG_DEBUG("WebRTC stack initialized before first use");
+    }
+    mPcConstraints.SetMandatoryReceiveAudio(true);
+    mPcConstraints.SetMandatoryReceiveVideo(true);
+    mPcConstraints.AddOptional(webrtc::MediaConstraintsInterface::kEnableDtlsSrtp, true);
+
+  //preload ice servers to make calls faster
+    initInputDevices();
+}
+
+promise::Promise<void> RtcModule::init(unsigned gelbTimeout)
+{
+    auto wptr = weakHandle();
+    return updateIceServers(gelbTimeout)
+    .fail([](const promise::Error& err)
+    {
+        RTCM_LOG_ERROR("updateIceServers failed, and it shouldn't. Error: %s", err.what());
+        assert(false);
+    })
+    .then([this, wptr]()
+    {
+        if (wptr.deleted())
+            return;
+        mClient.chatd->setRtcHandler(this);
+    });
+}
+
+IRtcModule* create(karere::Client &client, IGlobalHandler &handler, IRtcCrypto* crypto, const char* iceServers)
+{
+    return new RtcModule(client, handler, crypto, iceServers);
+}
+promise::Promise<void> RtcModule::updateIceServers(unsigned timeout)
+{
+    auto wptr = weakHandle();
+    return mTurnServerProvider.getServers(timeout)
+    .then([wptr, this](ServerList<TurnServerInfo>* servers)
+    {
+        if (wptr.deleted())
+            return;
+        setIceServers(*servers);
+    });
+}
+
+template <class T>
+T RtcModule::random() const
+{
+    T result;
+    crypto().random((char*)&result, sizeof(result));
+    return result;
+}
+template <class T>
+void RtcModule::random(T& result) const
+{
+    return crypto().random((char*)&result, sizeof(T));
+}
+
+void RtcModule::initInputDevices()
+{
+    auto& devices = mDeviceManager.inputDevices();
+    if (!devices.audio.empty())
+        selectAudioInDevice(devices.audio[0].name);
+    if (!devices.video.empty())
+        selectVideoInDevice(devices.video[0].name);
+    RTCM_LOG_INFO("Input devices on this system:");
+    for (const auto& dev: devices.audio)
+        RTCM_LOG_INFO("\tAudio: %s [id=%s]", dev.name.c_str(), dev.id.c_str());
+    for (const auto& dev: devices.video)
+        RTCM_LOG_INFO("\tVideo: %s [id=%s]", dev.name.c_str(), dev.id.c_str());
+}
+const cricket::Device* RtcModule::getDevice(const string& name, const artc::DeviceList& devices)
+{
+    for (size_t i=0; i<devices.size(); i++)
+    {
+        auto device = &devices[i];
+        if (device->name == name)
+            return device;
+    }
+    return nullptr;
+}
+
+bool RtcModule::selectDevice(const std::string& devname,
+            const artc::DeviceList& devices, string& selected)
+{
+    if (devices.empty())
+    {
+        selected.clear();
+        return devname.empty();
```

Comment:
Shouldn't it return false here, regardless the device name was empty (attempt to select default device), since no device has been selected by this function.